### PR TITLE
OHRM5X 28 : Admin - Qualification - Language screen [BE & FE)

### PR DESCRIPTION
**Pull request by ireshe**
- [x] added Languages API
- [x] added Languages Screens
- [x] remove comments from Educations API and Services

### CODE REVIEW

File: symfony/plugins/orangehrmAdminPlugin/entity/Language.php [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+ *
+ * You should have received a copy of the GNU General Public License along with this program;
+ * if not, write to the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA  02110-1301, USA
+ */
+
+namespace OrangeHRM\Entity;
+
+use Doctrine\Common\Collections\ArrayCollection;
+use Doctrine\Common\Collections\Collection;
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/entity/Language.php:23]: remove `use Doctrine\Common\Collections\ArrayCollection;` and `use Doctrine\Common\Collections\Collection;`

File: symfony/plugins/orangehrmAdminPlugin/test/Dao/LanguageDaoTest.php [hunk 2]
```diff
@@ -16,126 +16,110 @@
...
-		$this->fixture = sfConfig::get('sf_plugins_dir') . '/orangehrmAdminPlugin/test/fixtures/LanguageDao.yml';
-		TestDataService::populate($this->fixture);
-	}
+    /**
+     * Set up method
+     */
+    protected function setUp(): void
+    {
+        $this->languageDao = new LanguageDao();
+        $this->fixture = Config::get('ohrm_plugins_dir') . '/orangehrmAdminPlugin/test/fixtures/LanguageDao.yml';
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/test/Dao/LanguageDaoTest.php:44]: `Config::get('ohrm_plugins_dir')` ==> `Config::get(Config::PLUGINS_DIR)`, find and replace other place as well

File: symfony/plugins/orangehrmAdminPlugin/test/Dao/LanguageDaoTest.php [hunk 3]
```diff
@@ -16,126 +16,110 @@
...
-    
-    public function testDeleteLanguages() {
-        
-        $result = $this->languageDao->deleteLanguages(array(1, 2));
-        
+
+    public function testDeleteLanguages(): void
+    {
+        $toTobedeletedIds = [1, 2];
+        $result = $this->languageDao->deleteLanguages($toTobedeletedIds);
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/test/Dao/LanguageDaoTest.php:86]: can you add test assertion with empty array

File: symfony/plugins/orangehrmAdminPlugin/test/Service/LanguageServiceTest.php [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    private LanguageService $languageService;
+    private string $fixture;
+
+    /**
+     * Set up method
+     */
+    protected function setUp(): void
+    {
+        $this->languageService = new LanguageService();
+        $this->fixture = Config::get('ohrm_plugins_dir') . '/orangehrmAdminPlugin/test/fixtures/LanguageDao.yml';
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/test/Service/LanguageServiceTest.php:46]: Config::get('ohrm_plugins_dir') should change

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: symfony/plugins/orangehrmAdminPlugin/test/Dao/LanguageDaoTest.php (hunk 3 above)
Comment (ireshe) [reply symfony/plugins/orangehrmAdminPlugin/test/Dao/LanguageDaoTest.php:86]:
```
$nullValue = [];
$result = $this->languageDao->deleteLanguages($nullValue);
$this->assertEquals(0, $result);
```
you mean something like this right ?

File: symfony/client/src/orangehrmAdminPlugin/pages/qualificationLanguage/EditQualificationLanguage.vue [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    );
+    return {
+      http,
+    };
+  },
+
+  data() {
+    return {
+      isLoading: false,
+      license: {
```
Comment (RajithaKumara) [inline symfony/client/src/orangehrmAdminPlugin/pages/qualificationLanguage/EditQualificationLanguage.vue:78]: is `license`  related to this

File: symfony/client/src/orangehrmAdminPlugin/pages/qualificationLanguage/SaveQualificationLanguage.vue [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+<script>
+import {navigate} from '@orangehrm/core/util/helper/navigation';
+import {APIService} from '@orangehrm/core/util/services/api.service';
+
+export default {
+  data() {
+    return {
+      isLoading: false,
+      license: {
```
Comment (RajithaKumara) [inline symfony/client/src/orangehrmAdminPlugin/pages/qualificationLanguage/SaveQualificationLanguage.vue:62]: `license` ??

File: symfony/plugins/orangehrmAdminPlugin/Api/LanguageAPI.php [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+     */
+    public function setLanguageService(LanguageService $languageService): void
+    {
+        $this->languageService = $languageService;
+    }
+
+    /**
+     * @return EndpointGetOneResult
+     * @throws RecordNotFoundException
+     * @throws DaoException
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Api/LanguageAPI.php:77]: check the IDE warning. remove `@throws DaoException`

File: symfony/plugins/orangehrmAdminPlugin/Api/LanguageAPI.php [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+        return new EndpointCreateResult(LanguageModel::class, $languages);
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getValidationRuleForCreate(): ParamRuleCollection
+    {
+        return new ParamRuleCollection(
+            new ParamRule(self::PARAMETER_NAME),
```
[the remark this answers is not in the document]
Comment (RajithaKumara) [reply symfony/plugins/orangehrmAdminPlugin/Api/LanguageAPI.php:152]: you didn't fix this?
please refer  https://github.com/orangehrm/orangehrm/blob/5.0/symfony/plugins/orangehrmPimPlugin/Api/EmployeeAPI.php#L212
to define validation rules, in each action

File: symfony/plugins/orangehrmAdminPlugin/Api/LanguageAPI.php [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+            ),
+            new ParamRule(self::PARAMETER_NAME),
+        );
+    }
+
+    /**
+     *
+     * @return EndpointDeleteResult
+     * @throws Exception
+     */
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Api/LanguageAPI.php:221]:
```php
    /**
     * @inheritDoc
     */
```

File: symfony/plugins/orangehrmAdminPlugin/Api/LicenseAPI.php [hunk 10]
```diff
@@ -172,7 +172,11 @@ public function saveLicense(): License
     public function getValidationRuleForCreate(): ParamRuleCollection
     {
         return new ParamRuleCollection(
-            new ParamRule(self::PARAMETER_NAME),
+            new ParamRule(
+                self::PARAMETER_NAME,
+                new Rule(Rules::STRING_TYPE),
+                new Rule(Rules::LENGTH, [null, 120]),
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Api/LicenseAPI.php:178]: wrong validation, `new Rule(Rules::LENGTH, [null, 100]),` check the schema

File: symfony/plugins/orangehrmAdminPlugin/Dao/LanguageDao.php [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+    }
+
+    /**
+     * @param LanguageSearchFilterParams $languageSearchFilterParams
+     * @return Paginator
+     */
+    public function getLanguageListPaginator(LanguageSearchFilterParams $languageSearchFilterParams
+    ): Paginator {
+        $q = $this->createQueryBuilder(Language::class, 'l');
+        if (!is_null($languageSearchFilterParams->getSortField())) {
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Dao/LanguageDao.php:106]: remove 
```php
        if (!is_null($languageSearchFilterParams->getSortField())) {
            $q->addOrderBy($languageSearchFilterParams->getSortField(), $languageSearchFilterParams->getSortOrder());
        }
        if (!empty($languageSearchFilterParams->getLimit())) {
            $q->setFirstResult($languageSearchFilterParams->getOffset())
                ->setMaxResults($languageSearchFilterParams->getLimit());
        }
```
and follow this pattern
```php
$this->setSortingAndPaginationParams($q, $languageSearchFilterParams);
```
https://github.com/orangehrm/orangehrm/blob/5.0/symfony/plugins/orangehrmPimPlugin/Dao/EmployeeDao.php#L68

File: symfony/plugins/orangehrmAdminPlugin/entity/Language.php (hunk 1 above)
Comment (RajithaKumara) [reply symfony/plugins/orangehrmAdminPlugin/entity/Language.php:22]: `use Doctrine\Common\Collections\Collection;` also not using, remove it

File: symfony/plugins/orangehrmAdminPlugin/test/Dao/LanguageDaoTest.php [hunk 12]
```diff
@@ -16,126 +16,114 @@
...
         $this->assertEquals(2, $result);
-        $this->assertEquals(1, count($this->languageDao->getLanguageList()));       
-        
+
+        $nullValue = [];
+        $result = $this->languageDao->deleteLanguages($nullValue);
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/test/Dao/LanguageDaoTest.php:90]:
```php
$result = $this->languageDao->deleteLanguages([]);
```

File: symfony/plugins/orangehrmAdminPlugin/test/Service/LicenseServiceTest.php [hunk 13]
```diff
@@ -96,9 +96,7 @@ public function testGetLicenseByName(): void
     protected function setUp(): void
     {
         $this->licenseService = new LicenseService();
-        $this->fixture = Config::get(
-                'ohrm_plugins_dir'
-            ) . '/orangehrmAdminPlugin/test/fixtures/LicenseDao.yml';
+        $this->fixture = Config::get('ohrm_plugins_dir') . '/orangehrmAdminPlugin/test/fixtures/LicenseDao.yml';
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/test/Service/LicenseServiceTest.php:99]: `Config::get('ohrm_plugins_dir')`  ==> `Config::get(Config::PLUGINS_DIR)`

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: symfony/plugins/orangehrmAdminPlugin/Api/LicenseAPI.php (hunk 10 above)
Comment (samanthajayasinghe) [reply symfony/plugins/orangehrmAdminPlugin/Api/LicenseAPI.php:178]: @RajithaKumara any chance define this char length as class const?

File: symfony/client/src/orangehrmAdminPlugin/pages/qualificationLanguage/EditQualificationLanguage.vue (hunk 5 above)
Comment (Super-Chama) [reply symfony/client/src/orangehrmAdminPlugin/pages/qualificationLanguage/EditQualificationLanguage.vue:78]: `v-model="license.name"` should be language

File: symfony/plugins/orangehrmAdminPlugin/Api/LicenseAPI.php (hunk 10 above)
Comment (RajithaKumara) [reply symfony/plugins/orangehrmAdminPlugin/Api/LicenseAPI.php:178]:
```php
    public const PARAM_RULE_NAME_MAX_LENGTH = 100;
    // ...

            new ParamRule(
                self::PARAMETER_NAME,
                new Rule(Rules::STRING_TYPE),
                new Rule(Rules::LENGTH, [null, self::PARAM_RULE_NAME_MAX_LENGTH]),
            ),

```

how about this?

File: symfony/plugins/orangehrmAdminPlugin/Api/LanguageAPI.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (ireshe) [reply symfony/plugins/orangehrmAdminPlugin/Api/LanguageAPI.php:152]: okay I updated on all education, license and languages, is it okay add a rule for the common params as well ? @RajithaKumara 

`return new ParamRuleCollection(
            new ParamRule(CommonParams::PARAMETER_ID,
                new Rule(Rules::INT_TYPE),
                new Rule(Rules::LENGTH,[null,11]),
            ),
        );`

File: symfony/client/src/orangehrmAdminPlugin/pages/qualificationLanguage/EditQualificationLanguage.vue (hunk 5 above)
Comment (ireshe) [reply symfony/client/src/orangehrmAdminPlugin/pages/qualificationLanguage/EditQualificationLanguage.vue:78]: yeah sorry change that.

File: symfony/client/src/orangehrmAdminPlugin/pages/qualificationLanguage/SaveQualificationLanguage.vue (hunk 6 above)
Comment (ireshe) [reply symfony/client/src/orangehrmAdminPlugin/pages/qualificationLanguage/SaveQualificationLanguage.vue:62]: change to language

File: symfony/plugins/orangehrmAdminPlugin/Api/LanguageAPI.php (hunk 9 above)
Comment (ireshe) [reply symfony/plugins/orangehrmAdminPlugin/Api/LanguageAPI.php:227]: @RajithaKumara I don't understand this ? should I remove it, it was auto generated. So i kept as it is
